File: app/src/main/java/knf/kuma/directory/DirectoryPageFragment.java
package knf.kuma.directory;

import android.os.Bundle;
import android.preference.PreferenceManager;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ProgressBar;

import androidx.annotation.LayoutRes;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.annotation.UiThread;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.ViewModelProviders;
import androidx.paging.PagedList;
import androidx.recyclerview.widget.RecyclerView;
import butterknife.BindView;
import butterknife.ButterKnife;
import knf.kuma.BottomFragment;
import knf.kuma.R;
import knf.kuma.pojos.AnimeObject;

public class DirectoryPageFragment extends BottomFragment {
    @BindView(R.id.recycler)
    RecyclerView recyclerView;
    @BindView(R.id.progress)
    ProgressBar progress;
    private RecyclerView.LayoutManager manager;
    private DirectorypageAdapter adapter;
    private boolean isFirst = true;
    private boolean listUpdated = false;

    public DirectoryPageFragment() {
    }

    public static DirectoryPageFragment get(DirType type) {
        Bundle bundle = new Bundle();
        bundle.putInt("type", type.value);
        DirectoryPageFragment fragment = new DirectoryPageFragment();
        fragment.setArguments(bundle);
        return fragment;
    }

    @Override
    public void onActivityCreated(@Nullable Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);
        DirectoryViewModel model = ViewModelProviders.of(getActivity()).get(DirectoryViewModel.class);
        adapter = new DirectorypageAdapter(this);
        adapter.registerAdapterDataObserver(new RecyclerView.AdapterDataObserver() {
            @Override
            public void onItemRangeInserted(int positionStart, int itemCount) {
                super.onItemRangeInserted(positionStart, itemCount);
                if (positionStart == 0)
                    scrollTop();
            }

            @Override
            public void onItemRangeMoved(int fromPosition, int toPosition, int itemCount) {
                super.onItemRangeMoved(fromPosition, toPosition, itemCount);
                if (toPosition == 0)
                    scrollTop();
            }
        });
        getLiveData(model).observe(this, animeObjects -> {
            hideProgress();
            adapter.submitList(animeObjects);
            makeAnimation();
        });
        recyclerView.setAdapter(adapter);
    }

    private LiveData<PagedList<AnimeObject>> getLiveData(DirectoryViewModel model) {
        switch (getArguments().getInt("type", 0)) {
            default:
            case 0:
                return model.getAnimes(getContext());
            case 1:
                return model.getOvas(getContext());
            case 2:
                return model.getMovies(getContext());
        }
    }

    public void onChangeOrder() {
        if (getActivity() != null && adapter != null && recyclerView != null) {
            getLiveData(ViewModelProviders.of(getActivity()).get(DirectoryViewModel.class)).observe(this, animeObjects -> {
                hideProgress();
                listUpdated = true;
                adapter.submitList(animeObjects);
                makeAnimation();
                scrollTop();
            });
        }
    }

    private void hideProgress() {
        progress.post(() -> progress.setVisibility(View.GONE));
    }

    private void makeAnimation() {
        if (isFirst) {
            recyclerView.scheduleLayoutAnimation();
            isFirst = false;
        }
    }

    @UiThread
    private void scrollTop() {
        try {
            recyclerView.smoothScrollToPosition(0);
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        View view = inflater.inflate(getLayout(), container, false);
        ButterKnife.bind(this, view);
        manager = recyclerView.getLayoutManager();
        recyclerView.setLayoutManager(manager);
        isFirst = true;
        return view;
    }

    @LayoutRes
    private int getLayout() {
        if (PreferenceManager.getDefaultSharedPreferences(getContext()).getString("lay_type", "0").equals("0")) {
            return R.layout.recycler_dir;
        } else {
            return R.layout.recycler_dir_grid;
        }
    }

    @Override
    public void onReselect() {
        if (recyclerView != null && manager != null)
            manager.smoothScrollToPosition(recyclerView, null, 0);
    }

    enum DirType {
        ANIMES(0),
        OVAS(1),
        MOVIES(2);
        public int value;

        DirType(int value) {
            this.value = value;
        }
    }
}

File: app/src/main/java/knf/kuma/directory/DirectoryPageFragment.kt
package knf.kuma.directory

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ProgressBar
import androidx.annotation.LayoutRes
import androidx.annotation.UiThread
import androidx.fragment.app.viewModels
import androidx.lifecycle.LiveData
import androidx.lifecycle.Observer
import androidx.paging.PagedList
import androidx.recyclerview.widget.RecyclerView
import knf.kuma.BottomFragment
import knf.kuma.R
import knf.kuma.commons.PrefsUtil
import knf.kuma.commons.verifyManager
import org.jetbrains.anko.find

class DirectoryPageFragment : BottomFragment() {
    lateinit var recyclerView: RecyclerView
    lateinit var progress: ProgressBar
    private var manager: RecyclerView.LayoutManager? = null
    private var adapter: DirectoryPageAdapter? = null
    private var isFirst = true
    private var waitingScroll = false
    private var listUpdated = false
    private val model: DirectoryViewModel by viewModels()

    private lateinit var liveData: LiveData<PagedList<DirObject>>
    private lateinit var observer: Observer<PagedList<DirObject>>

    private val layout: Int
        @LayoutRes
        get() = if (PrefsUtil.layType == "0") {
            R.layout.recycler_dir
        } else {
            R.layout.recycler_dir_grid
        }

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)
        activity?.let {
            observeLiveData(model, Observer { animeObjects ->
                hideProgress()
                adapter?.submitList(animeObjects)
                makeAnimation()
            })
        }
    }

    private fun observeLiveData(model: DirectoryViewModel, newObserver: Observer<PagedList<DirObject>>) {
        if (::liveData.isInitialized && ::observer.isInitialized)
            liveData.removeObserver(observer)
        liveData = when (arguments?.getInt("type", 0) ?: 0) {
            1 -> model.getOvas()
            2 -> model.getMovies()
            else -> model.getAnimes()
        }
        liveData.observe(this, newObserver.also { observer = newObserver })
    }

    fun onChangeOrder() {
        activity?.let {
            waitingScroll = true
            adapter?.submitList(null)
            showProgress()
            observeLiveData(model, Observer { animeObjects ->
                hideProgress()
                listUpdated = true
                adapter?.submitList(animeObjects)
                makeAnimation()
            })
        }
    }

    private fun hideProgress() {
        progress.post { progress.visibility = View.GONE }
    }

    private fun showProgress() {
        isFirst = true
        progress.post { progress.visibility = View.VISIBLE }
    }

    private fun makeAnimation() {
        if (isFirst) {
            recyclerView.scheduleLayoutAnimation()
            isFirst = false
        }
    }

    @UiThread
    private fun scrollTop() {
        try {
            recyclerView.smoothScrollToPosition(0)
        } catch (e: Exception) {
            e.printStackTrace()
        }

    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        val view = inflater.inflate(layout, container, false)
        recyclerView = view.find(R.id.recycler)
        progress = view.find(R.id.progress)
        return view
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        manager = recyclerView.layoutManager
        recyclerView.layoutManager = manager
        adapter = DirectoryPageAdapter(this)
        adapter?.registerAdapterDataObserver(object : RecyclerView.AdapterDataObserver() {
            override fun onItemRangeInserted(positionStart: Int, itemCount: Int) {
                super.onItemRangeInserted(positionStart, itemCount)
                if (positionStart == 0 && waitingScroll) {
                    scrollTop()
                    waitingScroll = false
                }
            }

            override fun onItemRangeMoved(fromPosition: Int, toPosition: Int, itemCount: Int) {
                super.onItemRangeMoved(fromPosition, toPosition, itemCount)
                if (toPosition == 0 && waitingScroll) {
                    scrollTop()
                    waitingScroll = false
                }
            }
        })
        recyclerView.verifyManager()
        recyclerView.adapter = adapter
        isFirst = true
    }

    override fun onReselect() {
        manager?.smoothScrollToPosition(recyclerView, null, 0)
    }

    enum class DirType(var value: Int) {
        ANIMES(0),
        OVAS(1),
        MOVIES(2)
    }

    companion object {

        operator fun get(type: DirType): DirectoryPageFragment {
            val bundle = Bundle()
            bundle.putInt("type", type.value)
            val fragment = DirectoryPageFragment()
            fragment.arguments = bundle
            return fragment
        }
    }
}
